wlr_keyboard.h & wlr_keyboard_group.h files

MattBrooks95 opened the pull request:
I spent a lot of time trying to figure out how to get the Haskell library `xkbcommon` to build so that I could use its type definitions but no luck. I ended up making the types in wlhs just be opague pointers. (Is that the word? It's just a ptr to an empty data declaration).

If, in order to make a compositor people actually need real peek and poke instances that could be bad.

Similarly, I could not find where the `wl_keyboard_key_state` enum was declared. I can't find the enum declaration in the Wayland library code. I just looked at the names and values of the enum in Wayland's "test/data/example_client.h" and wrote the patterns by hand.

Also, there were some array types in the wlr_keyboard struct, so I just used a dummy type until the macro can support array declarations.

MattBrooks95 committed:
wlr_keyboard.h file

keyboard_modifiers

got started on keyboard, realized that I need to do keyboard_group first

start keyboard_group

there is an import cycle here in the C

and cabal is complaining

today I learned how to do import cycles in Haskell

saving source changes, can't get it to build though

start writing xkb structs

for now all we can do is the bare minimum to have the typing information
keyboard related structs from wlroots

keyboard code switched to templates, still need to get it to build

a lot of consts

it won't let me import these header files

nothing seems to go right

give up on doing the full xbkcommon structs

there's no end to these nested dependencies, and most of them can't be
imported from libxkbcommon, so I guess on the Haskell side
an untyped pointer will have to do

building with the struct macro

some more fields for wlr_keyboard

going to skip doing the array fields

more progress on keyboard

foreign calls for keyboard

MattBrooks95 commented on src/WLR/Types/Keyboard.hsc:
```diff
@@ -0,0 +1,169 @@
+
+type Wlr_keyboard_modifier = CInt
+-- enum wlr_keyboard_modifier {
+-- WLR_MODIFIER_SHIFT = 1 << 0,
```
I'm pretty sure these bit shifts are just declaring the value to be powers of two, instead of writing haskell code to calculate that I just wrote in the powers of 2 by hand

bradrn replied:
Yep, that looks correct to me.

bradrn commented:
> I ended up making the types in wlhs just be opague pointers.

That seems like a reasonable approach for now. I’m sure we will eventually need to use some things from `xkbcommon`, but we can deal with that problem when we get to it.

> (Is that the word? It's just a ptr to an empty data declaration).

‘Opaque’, yes. (I presume ‘opague’ was just a typo.)

bradrn committed:
Fix typo in README

dylan-hachmann commented on src/WLR/Types/Keyboard.hsc:
```diff
@@ -0,0 +1,169 @@
+}}
+
+-- cannot import these types from libxcommon because of their fields which have
+-- internal types that aren't exported
```
What do you mean by this? Regardless, looks like there's plenty of nested structs in each of those, so this is work that's best put off for another PR.

One request though; could you capitalize these as `XKB_keymap` and `XKB_state`? I think it's best to keep the naming consistent with what we've got already.

MattBrooks95 replied:
Yes I will caps the prefix.

About the imports though, I don't think we can use hsc features to create the bindings for things that are internal to xkbcommon. As far as I can tell they cannot be imported in any way, we'd have to write out something by hand that lines up with what they have perfectly.

I wonder if the fact that they're not exported means that we won't ever need to peek or poke them, and we just need to be able to pass them through to other C functions.

dylan-hachmann replied:
> things that are internal to xkbcommon

This is what I'm confused about. Could you give a few examples?

MattBrooks95 replied:
For example, I cannot find importable declarations of `struct xkb_keymap {`

There is a definition in `src/keymap.h` in the libxkbcommon repo. In that file they have this section:
```
 /* Don't use compat names in internal code. */
#define _XKBCOMMON_COMPAT_H
#include "xkbcommon/xkbcommon.h"
```
that implies that the file is 'internal code'

When I tried adding xkbcommon to cabal and including things from that file, it fails. I think they are hiding struct definitions that they don't want to be a part of the public API

MattBrooks95 replied:
On another branch I started looking at `wlr_seat`, and `wlr_seat` has a field for a `wl_global` that I cannot find.

[some structs are meant to be opaque](https://bugaevc.gitbooks.io/writing-wayland-clients/content/black-square/the-wayland-client-library.html)
> To keep track of objects, wayland-client stores some metadata about them in appropriately named opaque structures. You always use the structure instances via pointers and never care what's inside. You interact with a structure instance as if it was the Wayland object it represents.

dylan-hachmann replied:
> that implies that the file is 'internal code'

There are some functions that are declared in `keymap.h` which are local, but I assume that those are just the functions defined in `keymap-priv.c`. It looks like all of the functions in `keymap.c` are marked with `XKB_EXPORT`, so those should be fine to write bindings to.

From the initial comment I was asking about:

> internal types that aren't exported

This is what I was particularly confused about. While some functions may have hidden visibility, I don't think that this should affect any of our mappings to structs/enums, since those were never symbols to be exported in the first place.

Regarding the opaque structures, that's probably a good thing. Less work for us to do :)

dylan-hachmann commented on src/WLR/Types/KeyboardGroup.hsc:
```diff
@@ -0,0 +1,64 @@
+    wlr_keyboard_group_from_wlr_keyboard :: Ptr WLR_keyboard -> Ptr WLR_keyboard_group
+
+-- TODO I looked at the C code for this function and it looks pure, but I'm not sure if it actually is
+-- put it in IO for now
```
Hmm, I'm wondering if there's any issues with using the IO monad even when it's unnecessary. If there's nothing wrong with it then maybe it's best to always use it, so that we have a consistent interface between each function.

MattBrooks95 replied:
I thought when I read the FFI section of Real World Haskell, it made some distinction between pure foreign functions and non-pure foreign functions. If we're not sure if it's pure or it, it may be best to just default to putting it in the IO monad.

bradrn replied:
Personally, I’ve been putting everything in `IO`, and I strongly prefer doing that by default. If it really becomes a problem we can always go back and remove the `IO` where it isn’t needed.

dylan-hachmann commented on src/WLR/Types/Keyboard.hsc:
```diff
@@ -0,0 +1,169 @@
+    wlr_keyboard_from_input_device :: Ptr WLR_input_device -> IO (Ptr WLR_keyboard)
+
+foreign import capi "wlr/types/wlr_keyboard.h wlr_keyboard_set_keymap"
+    wlr_keyboard_set_keymap :: Ptr wLR_keyboard -> Ptr Xkb_keymap -> CBool
```
typo

MattBrooks95 replied:
I'm not sure why or how this built :thinking: thank you

dylan-hachmann commented:
> Similarly, I could not find where the wl_keyboard_key_state enum was declared. I can't find the enum declaration in the Wayland library code

It comes from `wayland-server-protocol.h`, which seems a be a generated header (you can see that in Wayland's `meson.build`). A bit annoying to have to work with generated headers, but oh well.

MattBrooks95 committed:
capitalizations and typo corrections

MattBrooks95 committed:
this does actually exist, just not in the wayland source

going forward, I need to check the headers that are actually in the
build result of wayland

on nix this meant I had to find it in the store :sweat:

thank goodness for `$ fd wayland-server-protocol.h`

MattBrooks95 committed:
use the CUInt type instead of Word32 directly

bradrn committed:
Merge branch 'keyboard_merge' of github.com:MattBrooks95/wlhs into MattBrooks95-keyboard_merge

bradrn committed:
Add new macros for array fields

bradrn committed:
Use array macro for WLR.Types.Keyboard

bradrn commented:
I’ve pushed to this PR my new array macros mentioned in #13, and have updated `WLR.Types.Keyboard` to use them. I haven’t tested the result, but it compiles and seems to expand to the correct code.

(One unexpected-but-neat consequence of the way I’ve implemented this: you can use any expression for the array length, including hsc2hs macros! This turns out to be rather useful here.)

bradrn commented:
Since no-one seems to have any strong objections to anything here, I’m just going to go ahead and merge this.
